apps/web/src/app/challenge/_components/fullscreen.tsx

```diff
@@ -1,22 +1,57 @@
 import { Tooltip, TooltipContent, TooltipTrigger } from '@repo/ui';
 import { Maximize2 } from '@repo/ui/icons';
-import { useLayoutSettingsStore } from './challenge-layout';
+import { Minimize2 } from 'lucide-react';
+import { create } from 'zustand';
+
+export const FS_SETTINGS = {
+  isFullscreen: false,
+};
+
+type FSSettings = typeof FS_SETTINGS;
+
+interface FSState {
+  fssettings: FSSettings;
+  updateFSSettings: (settings: Partial<FSSettings>) => void;
+}
+
+export const useFullscreenSettingsStore = create<FSState>()((set, get) => ({
+  fssettings: FS_SETTINGS,
+  updateFSSettings: (fssettings) => set({ fssettings: { ...get().fssettings, ...fssettings } }),
+}));
 
 export function FullscreenButton() {
-  const { settings, updateSettings } = useLayoutSettingsStore();
+  const { fssettings, updateFSSettings } = useFullscreenSettingsStore();
 
   const handleToggleFullscreen = () => {
-    updateSettings({ ...settings, isFullscreen: !settings.isFullscreen });
+    if (!document.fullscreenElement) {
+      document.documentElement.requestFullscreen();
+    } else if (document.exitFullscreen) {
+      document.exitFullscreen();
+    }
+    updateFSSettings({ ...fssettings, isFullscreen: !fssettings.isFullscreen });
+  };
+
+  // exiting fullscreen thru Esc key
+  const fullscreenchanged = () => {
+    if (!document.fullscreenElement) {
+      updateFSSettings({ ...fssettings, isFullscreen: false });
+    }
   };
 
+  document.addEventListener('fullscreenchange', fullscreenchanged);
```

Comment:
this needs to be inside a useEffect

https://github.com/bautistaaa/typehero/blob/main/apps/web/src/app/challenge/_components/challenge-layout.tsx#L43